File: WayzyyChat/Moderation/ActorSignals.swift
// Actor and behavioural signals over a 24-hour window: velocity, fan-out, reports and sub-threshold safety patterns.

import Foundation

struct ActorRisk {
    var messagesInWindow = 0
    var distinctConversations = 0
    var repeatTargetCount = 0
    var receivedReports = 0
    var blockEvents = 0
    var subThresholdSafetyHits = 0
    var escalating = false

    var composite: Double {
        var r = 0.0
        r += Swift.min(Double(receivedReports) * 0.30, 0.60)
        r += Swift.min(Double(blockEvents) * 0.20, 0.40)
        r += Swift.min(Double(subThresholdSafetyHits) * 0.12, 0.36)
        r += Swift.min(Double(distinctConversations) / 40.0, 0.20)
        r += Swift.min(Double(messagesInWindow) / 400.0, 0.10)
        return Swift.min(r, 1.0)
    }

    var isElevated: Bool { composite >= 0.35 }
}

final class ActorSignalStore {

    static let window: TimeInterval = 24 * 3_600
    static let patternThreshold = 3
    static let patternThresholdAfterReport = 2

    /// Storage only. Every threshold and all of the risk arithmetic stays in this type, so
    /// swapping in a shared backend cannot change what the numbers mean.
    private let backend: ActorSignalBackend

    init(backend: ActorSignalBackend = InMemoryActorSignalBackend()) {
        self.backend = backend
    }

    func observe(
        sender: String,
        conversation: String,
        safetyScore: Double,
        routedForSafety: Bool,
        acted: Bool,
        at now: Date = Date()
    ) {
        backend.mutate(sender: sender) { state in
            state.events.append(ActorSignalSnapshot.Event(
                at: now, conversation: conversation, safetyScore: safetyScore,
                routedForSafety: routedForSafety, acted: acted))
            state.prune(before: now.addingTimeInterval(-Self.window))
        }
    }

    func recordReport(against sender: String, at now: Date = Date()) {
        backend.mutate(sender: sender) { state in
            state.reports.append(now)
            state.prune(before: now.addingTimeInterval(-Self.window))
        }
    }

    func recordBlock(of sender: String, at now: Date = Date()) {
        backend.mutate(sender: sender) { state in
            state.blocks.append(now)
            state.prune(before: now.addingTimeInterval(-Self.window))
        }
    }

    /// Persist DB-backed violation counts so a replica that has not seen this sender still clamps.
    func notePlatformPriors(sender: String, count: Int) {
        guard count > 0 else { return }
        backend.mutate(sender: sender) { state in
            state.platformPriors = max(state.platformPriors, count)
        }
    }

    func platformPriors(for sender: String) -> Int {
        max(0, backend.snapshot(sender: sender)?.platformPriors ?? 0)
    }

    func risk(for sender: String, conversation: String, at now: Date = Date()) -> ActorRisk {
        guard var state = backend.snapshot(sender: sender) else { return ActorRisk() }
        // Prune on read as well as write: expired evidence must not count towards a threshold
        // just because the sender has been quiet since.
        state.prune(before: now.addingTimeInterval(-Self.window))

        var risk = ActorRisk()
        risk.messagesInWindow = state.events.count
        risk.distinctConversations = Set(state.events.map(\.conversation)).count
        risk.receivedReports = state.reports.count
        risk.blockEvents = state.blocks.count

        let here = state.events.filter { $0.conversation == conversation }
        risk.repeatTargetCount = here.count

        let subThreshold = here.filter { $0.routedForSafety && !$0.acted }
        risk.subThresholdSafetyHits = subThreshold.count

        let bar = risk.receivedReports > 0
            ? Self.patternThresholdAfterReport
            : Self.patternThreshold
        risk.escalating = subThreshold.count >= bar
        return risk
    }

    /// For startup reporting only: which storage is actually installed. A deployment that
    /// believes it moved actor state to a shared store and did not should be able to see that.
    var backendForDiagnostics: ActorSignalBackend { backend }

    func reset(sender: String) { backend.remove(sender: sender) }

    func resetAll() { backend.removeAll() }

    var trackedSenderCount: Int { backend.trackedSenderCount }
}

File: react-native/src/ActorSignals.ts
export interface ActorSignalEvent {
    at: Date;
    conversation: string;
    safetyScore: number;
    routedForSafety: boolean;
    acted: boolean;
}

export class ActorSignalSnapshot {
    events: ActorSignalEvent[] = [];
    reports: Date[] = [];
    blocks: Date[] = [];
    platformPriors: number = 0;

    prune(before: Date) {
        this.events = this.events.filter(e => e.at >= before);
        this.reports = this.reports.filter(r => r >= before);
        this.blocks = this.blocks.filter(b => b >= before);
    }

    get isEmpty(): boolean {
        return this.events.length === 0 && this.reports.length === 0 && this.blocks.length === 0 && this.platformPriors <= 0;
    }
}

export interface ActorSignalBackend {
    snapshot(sender: string): ActorSignalSnapshot | null;
    mutate(sender: string, body: (state: ActorSignalSnapshot) => void): void;
    remove(sender: string): void;
    removeAll(): void;
    get trackedSenderCount(): number;
}

export class InMemoryActorSignalBackend implements ActorSignalBackend {
    private senders: Map<string, ActorSignalSnapshot> = new Map();
    private readonly maxSenders: number;

    constructor(maxSenders: number = 20000) {
        this.maxSenders = maxSenders;
    }

    snapshot(sender: string): ActorSignalSnapshot | null {
        return this.senders.get(sender) || null;
    }

    mutate(sender: string, body: (state: ActorSignalSnapshot) => void): void {
        if (!this.senders.has(sender) && this.senders.size >= this.maxSenders) {
            this.evictOldest();
        }
        const state = this.senders.get(sender) || new ActorSignalSnapshot();
        body(state);
        
        if (state.isEmpty) {
            this.senders.delete(sender);
        } else {
            this.senders.set(sender, state);
        }
    }

    remove(sender: string): void {
        this.senders.delete(sender);
    }

    removeAll(): void {
        this.senders.clear();
    }

    get trackedSenderCount(): number {
        return this.senders.size;
    }

    private evictOldest(): void {
        let oldestSender: string | null = null;
        let oldestTime = new Date(8640000000000000); // Max date

        for (const [sender, state] of this.senders.entries()) {
            const lastEvent = state.events.length > 0 ? state.events[state.events.length - 1] : null;
            const time = lastEvent ? lastEvent.at : new Date(0);
            if (time < oldestTime) {
                oldestTime = time;
                oldestSender = sender;
            }
        }

        if (oldestSender) {
            this.senders.delete(oldestSender);
        }
    }
}

export class ActorRisk {
    messagesInWindow = 0;
    distinctConversations = 0;
    repeatTargetCount = 0;
    receivedReports = 0;
    blockEvents = 0;
    subThresholdSafetyHits = 0;
    escalating = false;

    get composite(): number {
        let r = 0.0;
        r += Math.min(this.receivedReports * 0.30, 0.60);
        r += Math.min(this.blockEvents * 0.20, 0.40);
        r += Math.min(this.subThresholdSafetyHits * 0.12, 0.36);
        r += Math.min(this.distinctConversations / 40.0, 0.20);
        r += Math.min(this.messagesInWindow / 400.0, 0.10);
        return Math.min(r, 1.0);
    }

    get isElevated(): boolean {
        return this.composite >= 0.35;
    }
}

export class ActorSignalStore {
    static readonly window: number = 24 * 3600 * 1000; // ms
    static readonly patternThreshold = 3;
    static readonly patternThresholdAfterReport = 2;

    private backend: ActorSignalBackend;

    constructor(backend: ActorSignalBackend = new InMemoryActorSignalBackend()) {
        this.backend = backend;
    }

    observe(
        sender: string,
        conversation: string,
        safetyScore: number,
        routedForSafety: boolean,
        acted: boolean,
        now: Date = new Date()
    ): void {
        this.backend.mutate(sender, state => {
            state.events.push({
                at: now,
                conversation,
                safetyScore,
                routedForSafety,
                acted
            });
            state.prune(new Date(now.getTime() - ActorSignalStore.window));
        });
    }

    recordReport(sender: string, now: Date = new Date()): void {
        this.backend.mutate(sender, state => {
            state.reports.push(now);
            state.prune(new Date(now.getTime() - ActorSignalStore.window));
        });
    }

    recordBlock(sender: string, now: Date = new Date()): void {
        this.backend.mutate(sender, state => {
            state.blocks.push(now);
            state.prune(new Date(now.getTime() - ActorSignalStore.window));
        });
    }

    notePlatformPriors(sender: string, count: number): void {
        if (count <= 0) return;
        this.backend.mutate(sender, state => {
            state.platformPriors = Math.max(state.platformPriors, count);
        });
    }

    platformPriors(sender: string): number {
        const state = this.backend.snapshot(sender);
        return Math.max(0, state ? state.platformPriors : 0);
    }

    risk(sender: string, conversation: string, now: Date = new Date()): ActorRisk {
        const state = this.backend.snapshot(sender);
        const risk = new ActorRisk();
        if (!state) return risk;

        state.prune(new Date(now.getTime() - ActorSignalStore.window));

        risk.messagesInWindow = state.events.length;
        risk.distinctConversations = new Set(state.events.map(e => e.conversation)).size;
        risk.receivedReports = state.reports.length;
        risk.blockEvents = state.blocks.length;

        const here = state.events.filter(e => e.conversation === conversation);
        risk.repeatTargetCount = here.length;

        const subThreshold = here.filter(e => e.routedForSafety && !e.acted);
        risk.subThresholdSafetyHits = subThreshold.length;

        const bar = risk.receivedReports > 0
            ? ActorSignalStore.patternThresholdAfterReport
            : ActorSignalStore.patternThreshold;
        
        risk.escalating = subThreshold.length >= bar;
        return risk;
    }

    reset(sender: string): void {
        this.backend.remove(sender);
    }

    resetAll(): void {
        this.backend.removeAll();
    }

    get trackedSenderCount(): number {
        return this.backend.trackedSenderCount;
    }
}
